src/psm/src/node.h

```diff
@@ -1,108 +1,167 @@
+// POSSIBILITY OF SUCH DAMAGE.
 
-Copyright (c) 2020, The Regents of the University of Minnesota
+#pragma once
 
-All rights reserved.
+#include <map>
+#include <memory>
+#include <set>
+#include <string>
+#include <vector>
 
-Redistribution and use in source and binary forms, with or without
-modification, are permitted provided that the following conditions are met:
+#include "connection.h"
+#include "odb/geom.h"
 
-* Redistributions of source code must retain the above copyright notice, this
-  list of conditions and the following disclaimer.
+namespace odb {
+class dbTechLayer;
+class dbBPin;
+class dbITerm;
+}  // namespace odb
 
-* Redistributions in binary form must reproduce the above copyright notice,
-  this list of conditions and the following disclaimer in the documentation
-  and/or other materials provided with the distribution.
+namespace utl {
+class Logger;
+}
 
-* Neither the name of the copyright holder nor the names of its
-  contributors may be used to endorse or promote products derived from
-  this software without specific prior written permission.
+namespace psm {
+class Shape;
 
-THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
-AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
-FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
-DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
-SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*/
+class Node
+{
+ public:
+  enum class NodeType
+  {
+    None,
+    Source,
+    ITerm,
+    BPin
+  };
 
-#pragma once
+  struct Compare
+  {
+    bool operator()(const Node* lhs, const Node* rhs) const
+    {
+      return lhs->compare(rhs);
+    }
+  };
 
-#include <map>
+  using NodeSet = std::set<Node*, Compare>;
 
-#include "odb/db.h"
-#include "utl/Logger.h"
+  Node(const odb::Point& pt, odb::dbTechLayer* layer);
+  ~Node() = default;
 
-namespace psm {
+  bool compare(const Node* other) const;
+  bool compare(const std::unique_ptr<Node>& other) const;
+
+  const odb::Point& getPoint() const { return pt_; };
+  odb::dbTechLayer* getLayer() const { return layer_; };
+
+  void print(utl::Logger* logger, const std::string& prefix = "") const;
+  virtual std::string describe(const std::string& prefix) const;
 
-using odb::dbInst;
-using odb::Point;
-using NodeIdx = int;
+  std::string getName() const;
+  std::string getTypeName() const;
 
-struct NodeEnclosure
+ protected:
+  virtual NodeType getType() const { return NodeType::None; }
+  double getDBUs() const;
```

Comment:
Do we expect each Node to have its own DBUs?  This seems rather low for this API.

Reply:
it's just a helper function to the layer to get the DBUs when printing debugging information. This is not meant for the public access.